files/default/neutron-ha-tool.py

```diff
@@ -249,13 +264,48 @@ def run(args):
 
 
 class RandomAgentPicker(object):
-    def __init__(self, agents):
+    def __init__(self, qclient, agents):
         self.agents = agents
 
     def pick(self):
         return random.choice(self.agents)
 
 
+class LeastBusyAgentPicker(object):
+    def __init__(self, qclient, agents):
+        self.now = datetime.datetime.now
+        self.cache_created_at = None
+        self.qclient = qclient
+        self.agents_by_id = dict((agent['id'], agent) for agent in agents)
+        self.router_count_per_agent_id = dict()
+        self.refresh_router_count_per_agent_id()
+
+    def refresh_router_count_per_agent_id(self):
+        LOG.info("Refreshing router count per agent cache")
+        self.router_count_per_agent_id = dict()
+        for agent_id in self.agents_by_id.keys():
+            self.router_count_per_agent_id[agent_id] = len(
+                list_routers_on_l3_agent(self.qclient, agent_id)
+            )
+        self.cache_created_at = self.now()
+
+    def cache_expired(self):
+        cache_life = self.now() - self.cache_created_at
+        return cache_life.total_seconds() > ROUTER_CACHE_MAX_AGE_SECONDS
+
+    def pick(self):
+        if self.cache_expired():
+            self.refresh_router_count_per_agent_id()
+
+        agent_id_number_of_routers = sorted(
+            self.router_count_per_agent_id.items(),
+            key=lambda x: (x[1], x[0])
+        )
+        agent_id = agent_id_number_of_routers[0][0]
```

Comment:
what happens here is an `IndexError` is thrown. Basically the same would happen if we used the original implementation with random:

```python
>>> import random
>>> random.choice([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/random.py", line 275, in choice
    return seq[int(self.random() * len(seq))]  # raises IndexError if seq is empty
IndexError: list index out of range
```

So I think let's leave it as it is?